### PR TITLE
Implement fromListRevN

### PR DESCRIPTION
**Pull request by rnjtranjan**

### CODE REVIEW

**Comment by rnjtranjan** [comment]
Fixes:https://github.com/composewell/streamly/issues/1632

File: core/src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs [hunk 1]
```diff
@@ -1849,6 +1849,24 @@ writeNWith alloc n = FL.take n (writeNWithUnsafe alloc n)
...
+        arr <- newArray (max n 0)
+        return $ FL.Partial (arr, n - 1)
+
+    step (arr, i) x =
+        if i < 0
+        then
+            return $ FL.Done arr
+        else do
+            putIndex i x arr
+            return $ FL.Partial (arr, i - 1)
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs:1868]: Can you write this in the same way as `writeN` has been written? Using an array in the state would thread around an additional variable which might cause trouble with optimization. Also the tuple in the state is not strict.

File: core/src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs [hunk 2]
```diff
@@ -1849,6 +1849,27 @@ writeNWith alloc n = FL.take n (writeNWithUnsafe alloc n)
 writeN :: forall m a. (MonadIO m, Storable a) => Int -> Fold m a (Array a)
 writeN = writeNWith newArray
 
+-- /Pre-release/
+{-# INLINE writeRevN #-}
+writeRevN :: forall m a. (MonadIO m, Storable a) => Int -> Fold m a (Array a)
+writeRevN n = Fold step initial (return .fromArrayUnsafe . fst)
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs:1855]: Add space after "."

File: core/src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs [hunk 3]
```diff
@@ -1849,6 +1849,27 @@ writeNWith alloc n = FL.take n (writeNWithUnsafe alloc n)
...
+    where
+
+    initial = do
+        arr <- newArray (max n 0)
+        return $ FL.Partial (toArrayUnsafe arr, n - 1)
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs:1860]: Do we need to store the count as we have start and end of the array? Having redundant/additional data in state causes performance issues. Also, we are not using a strict tuple here.

File: core/src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs [hunk 4]
```diff
@@ -1849,6 +1849,27 @@ writeNWith alloc n = FL.take n (writeNWithUnsafe alloc n)
...
+
+    step (arr@(ArrayUnsafe contents start end), i) x =
+        if i < 0
+        then
+            return $ FL.Done (fromArrayUnsafe arr)
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs:1865]: Bug: we are losing the element `x` here.

File: core/src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs [hunk 5]
```diff
@@ -1849,6 +1849,27 @@ writeNWith alloc n = FL.take n (writeNWithUnsafe alloc n)
...
+        else
+            do
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs:1867]: Collapse else and "do" on a single line.

File: core/src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs [hunk 6]
```diff
@@ -1849,6 +1849,27 @@ writeNWith alloc n = FL.take n (writeNWithUnsafe alloc n)
...
+            liftIO $ poke (PTR_INDEX(start, i, a)) x >> touch contents
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs:1868]: Bring "touch" on the next line.

**Review by harendra-kumar — CHANGES_REQUESTED** [review]

File: core/src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs (hunk 3 above)
Comment (rnjtranjan) [reply core/src/Streamly/Internal/Data/Array/Foreign/Mut/Type.hs]: this is the index where we poke the element in reverse order.

**Review by rnjtranjan — COMMENTED** [review]
looks good.

**Pull request merged**